sign/slhdsa/internal.go

```diff
@@ -0,0 +1,137 @@
+package slhdsa
+
+import "encoding/binary"
+
+// See FIPS 205 -- Section 9
+// SLH-DSA Internal Functions
+
+// See FIPS 205 -- Section 9.1 -- Algorithm 18.
+func slhKeyGenInternal(
+	p *params, skSeed, skPrf, pkSeed []byte,
+) (pub PublicKey, priv PrivateKey) {
+	s := p.NewStatePriv(skSeed, pkSeed)
+	defer s.Clear()
+
+	stack := p.NewStack(p.hPrime)
+	defer stack.Clear()
+
+	addr := p.NewAddress()
+	addr.SetLayerAddress(p.d - 1)
+	pkRoot := make([]byte, p.n)
+	s.xmssNodeIter(stack, pkRoot, 0, p.hPrime, addr)
+
+	pub.ParamID = p.id
+	pub.seed = pkSeed
+	pub.root = pkRoot
+
+	priv.ParamID = p.id
+	priv.prfKey = skPrf
+	priv.seed = skSeed
+	priv.publicKey = pub
+
+	return
+}
+
+func (p *params) parseMsg(
+	digest []byte,
+) (md []byte, idxTree [3]uint32, idxLeaf uint32) {
+	l1 := (p.k*p.a + 7) / 8
+	l2 := (p.h - p.h/p.d + 7) / 8
+	l3 := (p.h + 8*p.d - 1) / (8 * p.d)
+
+	c := cursor(digest)
+	md = c.Next(l1)
+	s2 := c.Next(l2)
+	s3 := c.Next(l3)
+
+	var b2 [12]byte
+	copy(b2[12-len(s2):], s2)
+	n2 := p.h - p.h/p.d
+	idxTree[0] = binary.BigEndian.Uint32(b2[8:]) & ((1 << n2) - 1)
+	n2 -= 32
+	idxTree[1] = binary.BigEndian.Uint32(b2[4:]) & ((1 << n2) - 1)
+	idxTree[2] = binary.BigEndian.Uint32(b2[0:])
+
+	var b3 [4]byte
+	copy(b3[4-len(s3):], s3)
+	mask32 := (uint32(1) << (p.h / p.d)) - 1
+	idxLeaf = mask32 & binary.BigEndian.Uint32(b3[0:])
+
+	return
+}
+
+// See FIPS 205 -- Section 9.2 -- Algorithm 19.
+func slhSignInternal(
+	p *params, sk *PrivateKey, message, addRand []byte,
+) ([]byte, error) {
+	sigBytes := make([]byte, p.SignatureSize())
+
+	var sig signature
+	curSig := cursor(sigBytes)
+	if !sig.fromBytes(p, &curSig) {
+		return nil, ErrSigParse
+	}
+
+	p.PRFMsg(sig.rnd, sk.prfKey, addRand, message)
+	digest := make([]byte, p.m)
+	p.HashMsg(digest, sig.rnd, message, &sk.publicKey)
+
+	md, idxTree, idxLeaf := p.parseMsg(digest)
+	addr := p.NewAddress()
+	addr.SetTreeAddress(idxTree)
+	addr.SetTypeAndClear(addressForsTree)
+	addr.SetKeyPairAddress(idxLeaf)
+
+	s := p.NewStatePriv(sk.seed, sk.publicKey.seed)
+	defer s.Clear()
+
+	s.forsSign(sig.forsSig, md, addr)
+	pkFors := s.forsPkFromSig(md, sig.forsSig, addr)
```

Comment:
Would it be possible to compute `pkFors` as a side effect of `forsSign`? Can save a few hashes if so. Similarly might be able to do this in the `htSign` function